File: iosApp/iosApp/view/characterlist/CharactersListScreen.swift
//
//  CharactersListScreen.swift
//  iosApp
//
//  Created by Ade Dyas  on 01/04/21.
//

import SwiftUI
import shared

struct CountriesListScreen: View {
    @EnvironmentObject var viewModel: AppViewModel
    
    var body: some View {
        let charactersListState = CharactersListStateProvider(stateProvider : viewModel.stateProvider).getCountriesListState()
        let events = CharactersListEvents(events : viewModel.events)
        if charactersListState.isLoading {
            LoadingElement()
        } else {
            NavigationView {
                List {
                    if charactersListState.charactersListItems.count == 0 {
                        HStack(spacing: 0) {
                            Text("empty list")
                        }
                    } else {
//                        Section(header: CountriesListHeader()) {
                            ForEach (charactersListState.charactersListItems, id: \.name) { item in
//                                NavigationLink(destination: CountryDetailScreen(countryName: item.name)) {
                                    CharactersListRow(
                                        item: item,
                                        favorite: charactersListState.favoriteCharacters[item.name] != nil,
                                        onFavoriteIconClick: { events.selectFavorite(country: item.name) }
                                    )
//                                }
                            }
//                        }
                    }
                }
                .navigationBarTitleDisplayMode(.inline)
                .toolbar {
                    ToolbarItem(placement: .principal) {
                        Text("My Rick Morty Characters List").font(.headline).foregroundColor(.white)
                    }
//                    ToolbarItemGroup(placement: .bottomBar) {
//                        CountriesListBottomBar(
//                            selectedItem : countriesListState.selectedMenuItem,
//                            onItemClick: { menuItem in events.selectMenuItem(menuItem: menuItem) }
//                        )
//                    }
                }
            }
            .navigationBarColor(backgroundColor: "6200EE".toUIColor(), tintColor: .white)
            .toolbarColor(backgroundColor: "6200EE".toUIColor(), tintColor: .white)
        }
    }
    
}

File: androidApp/src/main/java/id/adeds/androidapp/view/list/CharactersListScreen.kt
package id.adeds.androidapp.view.list


import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.*
import androidx.compose.runtime.Composable
import androidx.compose.foundation.lazy.items
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import id.adeds.androidapp.ui.element.LoadingElement
import id.adeds.shared.viewmodel.characterslist.CharactersListEvents
import id.adeds.shared.viewmodel.characterslist.CharactersListState

@Composable
fun CountriesListScreen(charactersListState: CharactersListState, events : CharactersListEvents, onListItemClick: (String) -> Unit) {

    Scaffold(
        topBar = {
            TopAppBar(title = {
                Text(
                    text = "My Rick Morty Characters List",
                    fontSize = 20.sp,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth()
                )
            })
        },
        content = { paddingValues ->
            if (charactersListState.isLoading) {
                LoadingElement()
            } else {
                if (charactersListState.charactersListItems.isEmpty()) {
                    Text(
                        text = "empty list",
                        style = MaterialTheme.typography.body1,
                        modifier = Modifier
                            .padding(top = 30.dp)
                            .fillMaxWidth(),
                        textAlign = TextAlign.Center,
                        fontSize = 18.sp
                    )
                } else {
                    LazyColumn(contentPadding = paddingValues) {
                        items(items = charactersListState.charactersListItems, itemContent = { item ->
                            CountriesListRow(
                                item = item,
                                favorite = charactersListState.favoriteCharacters.containsKey(item.name),
                                onItemClick = { onListItemClick(item.name) },
                                onFavoriteIconClick = { events.selectFavorite(item.name) })
                        })
                    }
                }
            }
        }/*,
        bottomBar = {
            CharactersListBottomBar(selectedItem = charactersListState.selectedMenuItem, onItemClick = { events.selectMenuItem(it) })
        }*/
    )
}